Sources/NIO/EventLoopFuture.swift

```diff
@@ -828,14 +828,41 @@ extension EventLoopFuture {
     ///     - eventLoop: The `EventLoop` on which the new `EventLoopFuture` callbacks will fire.
     /// - returns: A new `EventLoopFuture`.
     public static func andAll(_ futures: [EventLoopFuture<Void>], eventLoop: EventLoop) -> EventLoopFuture<Void> {
-        let p0: EventLoopPromise<Void> = eventLoop.newPromise()
+        let body = EventLoopFuture<Void>.reduce((), futures, eventLoop: eventLoop) {
+            (_, _) in
+        }
+        return body
+    }
+    
+    /// Returns a new `EventLoopFuture` that fires only when all the provided futures complete.
+    /// The new `EventLoopFuture` contains the result of reducing the `initialResult` with the
+    /// values of the `[EventLoopFuture<U>]`.
+    ///
+    /// returned `EventLoopFuture` will fail as soon as any of the futures fails: otherwise, it will succeed
+    /// only when all of them do.
+    ///
+    /// - parameters:
+    ///     - initialResult: An initial result to being the reduction.
+    ///     - futures: An array of `EventLoopFuture<Void>` to wait for.
+    ///     - eventLoop: The `EventLoop` on which the new `EventLoopFuture` callbacks will fire.
+    ///     - reducer: The bifunction used to produce partial results.
+    /// - returns: A new `EventLoopFuture` with the reduced value.
+    public static func reduce<U>(_ initialResult: T, _ futures: [EventLoopFuture<U>], eventLoop: EventLoop, reducer: @escaping (T, U) -> T) -> EventLoopFuture<T> {
+        let p0: EventLoopPromise<T> = eventLoop.newPromise()
         guard futures.count > 0 else {
-            p0.succeed(result: ())
+            p0.succeed(result: initialResult)
             return p0.futureResult
         }
-
-        let body: EventLoopFuture<Void> = futures.reduce(p0.futureResult, { (f1: EventLoopFuture<Void>, f2: EventLoopFuture<Void>) in f1.and(f2).map({ (_ : ((), ())) in }) })
-        p0.succeed(result: ())
+        
+        func nextPartialResult(_ arg: (T, U)) -> (T) {
+            let (partialResult, nextItem) = arg
+            return reducer(partialResult, nextItem)
+        }
+        
+        let body = futures.reduce(p0.futureResult) { (f1: EventLoopFuture<T>, f2: EventLoopFuture<U>) -> EventLoopFuture<T> in
+            f1.and(f2).map(nextPartialResult)
+        }
+        p0.succeed(result: initialResult)
```

Comment:
Do we want a `thenReduce` function that is very similar but where the reducer itself returns an `EventLoopFuture`? Not sure if we're wildly over-engineering with that case or not.

Reply:
The problem with the current approach of `and` followed by `map` is that `and` allocates `EventLoopPromise` and `map` allocates `EventLoopPromise` and `EventLoopFuture`. If we implement a `thenReduce(otherFuture:bifunction)` we could avoid the overhead of these allocations.